pkgs/stdenv/generic/check-meta.nix

```diff
@@ -217,6 +251,7 @@ let
     downloadPage = str;
     changelog = either (listOf str) str;
     license = either (listOf lib.types.attrs) (either lib.types.attrs str);
+    sourceProvenance = either (listOf lib.types.attrs) (either lib.types.attrs str);
```

Comment:
```suggestion
    sourceProvenance = either (listOf lib.types.attrs) lib.types.attrs;
```

Do we necessarily want to include the `str` option?

I am assuming this is only here for analogy with the `license` option, as the documentation already states that it should be an attribute set.

I am conjecturing that allowing `str` in licenses is only for a historical reason (although I do not know and have not checked!).

When allowing `str` outside of lists, why not allow it inside of lists as well?

If we opt to allow `str` outside of lists, we have to update the implementation, because building this configuration currently succeeds:

```nix
# google-chome/default.nix
{
  # ...
  sourceProvenance = "binaryNativeCode";
}
```

```nix
# configuration.nix
{
  environment.systemPackages = [ pkgs.google-chrome ];
  nixpkgs.config.allowUnfree = true;
  nixpkgs.config.allowNonSource = false;
}
```

I have reviewed the remainder of the code assuming that we forbid `str` `sourceTypes`.

Reply:
> I don't know.

@risicle, if you don't have a reason to want the `str` option here, please consider dropping it.

The point of this PR is to enable automated checking.  I can't help but imagine that some joker is going to write

```
sourceProvenance = "I dunno";
```

> leave this PR as it is (perhaps as a "relic") and for others to create their own competing PRs

It is much easier to add more options later than to remove options.

Reply:
> It is much easier to add more options later than to remove options.

My point is it might be better to not merge this PR _at all_, in favour of an improved one that you two appear to be discussing productively. My time is and attention is limited at the moment and I'm finding it hard to keep up with the discussion enough to do it justice.

Reply:
> My point is it might be better to not merge this PR _at all_, in favour of an improved one that you two appear to be discussing productively. My time is and attention is limited at the moment and I'm finding it hard to keep up with the discussion enough to do it justice.

@risicle, one of the reasons why I use github's "suggest changes" feature is to make it very very easy for you to integrate them.  Just two mouse clicks.

If you don't have time for that, you will need to close this PR and I will reopen it and take responsibility for carrying it across the finish line.

Github does not have any way to reassign pull requests.

Reply:
It is not just a matter of blindly accepting peoples' suggestions, but also understanding and concurring with the rationale.

Reply:
Ok so for this one, there's nothing stopping someone reinstating string-based values later if they object..